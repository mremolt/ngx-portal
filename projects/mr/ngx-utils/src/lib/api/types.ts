import { HttpErrorResponse, HttpRequest } from '@angular/common/http';
import { Action } from '@ngrx/store';

export interface Constructor<T> {
  new (...args: any[]): T;
}

export interface ApiTokenSelector {
  select(state: any): string;
}

export interface ApiTokenSelectorConstructor {
  new (): ApiTokenSelector;
}

export interface ApiStartAction extends Action {
  readonly payload?: any;
}

export interface ApiSuccessAction<T> extends Action {
  readonly payload: T;
}

export interface ApiErrorAction extends Action {
  readonly payload: HttpErrorResponse;
}

export type ApiActions<T> = Readonly<{
  start: Constructor<ApiStartAction>;
  success: Constructor<ApiSuccessAction<T>>;
  error: Constructor<ApiErrorAction>;
  complete?: Constructor<Action>;
}>;

export type ApiRequestConfig<T> = Readonly<{
  actions: ApiActions<T>;
  request: HttpRequest<T>;
}>;
