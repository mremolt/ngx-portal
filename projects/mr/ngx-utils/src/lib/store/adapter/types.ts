import { DeepReadonly } from 'utility-types';

export type NormalizedEntities<Entity extends object> = DeepReadonly<{
  collection: { [key: string]: Entity };
  [key: string]: { [key: string]: object };
}>;

export type CollectionState<Entity extends object> = DeepReadonly<{
  entities: NormalizedEntities<Entity>;
  result: string[];
  selected: string | null;
  loading: boolean;
  loaded: boolean;
  dirty: boolean;
  updatedAt: Date | null;
}>;
