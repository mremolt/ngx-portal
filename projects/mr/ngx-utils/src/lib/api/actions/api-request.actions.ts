import { HttpErrorResponse, HttpRequest } from '@angular/common/http';
import { Action } from '@ngrx/store';

import { ApiActions, ApiRequestConfig } from '../types';

export enum ApiRequestActionTypes {
  ApiRequest = '[API] Request',
  ApiError = '[API] Error',
}

export class ApiRequest<T = any> implements Action {
  public readonly type = ApiRequestActionTypes.ApiRequest;

  constructor(public payload: ApiRequestConfig<T>) {}
}

export class ApiError implements Action {
  public readonly type = ApiRequestActionTypes.ApiError;

  constructor(public payload: HttpErrorResponse) {}
}

export class ApiGetAction<T> extends ApiRequest {
  constructor(url: string, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('GET', url);
    super({ request, actions });
  }
}

export class ApiPostAction<T> extends ApiRequest {
  constructor(url: string, body: T | null, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('POST', url, body);
    super({ request, actions });
  }
}

export class ApiPutAction<T> extends ApiRequest {
  constructor(url: string, body: T | null, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('PUT', url, body);
    super({ request, actions });
  }
}

export class ApiDeleteAction<T> extends ApiRequest {
  constructor(url: string, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('DELETE', url);
    super({ request, actions });
  }
}
