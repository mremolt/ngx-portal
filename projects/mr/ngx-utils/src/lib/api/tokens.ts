import { InjectionToken } from '@angular/core';

import { ApiTokenSelector } from './types';

export const API_TOKEN_SELECTOR = new InjectionToken<ApiTokenSelector>('TOKEN_SELECTOR');
