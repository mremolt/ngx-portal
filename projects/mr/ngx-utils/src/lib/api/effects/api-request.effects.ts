import { HttpClient, HttpRequest, HttpResponse } from '@angular/common/http';
import { Inject, Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { Store, select } from '@ngrx/store';
import { EMPTY, concat, of } from 'rxjs';
import { catchError, filter, flatMap, map, share, take } from 'rxjs/operators';

import { APP_ENVIRONMENT } from '../../tokens';
import { Environment } from '../../types';
import { ApiError, ApiRequest, ApiRequestActionTypes } from '../actions/api-request.actions';
import { API_TOKEN_SELECTOR } from '../tokens';
import { ApiTokenSelector } from '../types';

export function buildApiHttpRequest<T>(
  baseRequest: HttpRequest<T>,
  env: Environment,
  token: string
) {
  return baseRequest.clone({
    url: `${env.apiUrl}/${baseRequest.url}`,
    setHeaders: { Authorization: `Bearer ${token}` },
  });
}

@Injectable()
export class ApiRequestEffects {
  @Effect()
  processApiRequest$ = this.actions$.pipe(
    ofType<ApiRequest>(ApiRequestActionTypes.ApiRequest),
    flatMap(action => {
      const baseRequest = action.payload.request;
      const actions = action.payload.actions;

      return this.store.pipe(
        select(this.tokenSelector.select),
        take(1),
        flatMap(token => {
          const req = buildApiHttpRequest(baseRequest, this.env, token);

          return concat(
            of(new actions.start()),
            this.http.request(req).pipe(
              filter<any>(res => res instanceof HttpResponse),
              map((res: HttpResponse<any>) => res.body),
              map(data => new actions.success(data)),
              catchError(e => of(new actions.error(e), new ApiError(e)))
            ),
            actions.complete ? of(new actions.complete()) : EMPTY
          );
        })
      );
    }),
    share()
  );

  constructor(
    private actions$: Actions,
    private http: HttpClient,
    private store: Store<any>,
    @Inject(API_TOKEN_SELECTOR) private tokenSelector: ApiTokenSelector,
    @Inject(APP_ENVIRONMENT) private env: Environment
  ) {}
}
