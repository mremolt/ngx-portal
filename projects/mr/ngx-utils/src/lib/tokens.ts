import { InjectionToken } from '@angular/core';

import { Environment } from './types';

export const APP_ENVIRONMENT = new InjectionToken<Environment>('APP_ENVIRONMENT');
