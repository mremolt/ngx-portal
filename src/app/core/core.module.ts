import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { APP_ENVIRONMENT, ApiModule, RouterModule } from '@mr/ngx-utils';
import { EffectsModule } from '@ngrx/effects';

import { environment } from '../../environments/environment';
import { AppEffects } from '../app.effects';
import { AuthModule } from '../auth/auth.module';
import { TokenSelector } from '../auth/reducers/auth.selectors';
import { ReducersModule } from '../reducers/reducers.module';

@NgModule({
  imports: [
    // Angular
    CommonModule,
    HttpClientModule,
    // ngrx
    ReducersModule,
    EffectsModule.forRoot([AppEffects]),
    // Features
    RouterModule,
    ApiModule.forRoot(TokenSelector, environment),
    AuthModule,
  ],
  declarations: [],
  providers: [
    {
      provide: APP_ENVIRONMENT,
      useValue: environment,
    },
  ],
})
export class CoreModule {}
