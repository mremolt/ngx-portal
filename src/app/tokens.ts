import { InjectionToken } from '@angular/core';

import { Environment } from '../environments/environment';

export const APP_ENVIRONMENT = new InjectionToken<Environment>('APP_ENVIRONMENT');
