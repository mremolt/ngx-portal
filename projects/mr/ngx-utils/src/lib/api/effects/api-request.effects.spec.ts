import { HttpErrorResponse, HttpRequest } from '@angular/common/http';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { TestBed } from '@angular/core/testing';
import { provideMockActions } from '@ngrx/effects/testing';
import { Action, StoreModule } from '@ngrx/store';
import { ReplaySubject } from 'rxjs';
import { bufferCount } from 'rxjs/operators';

import { APP_ENVIRONMENT } from '../../tokens';
import { ApiError, ApiRequest } from '../actions/api-request.actions';
import { API_TOKEN_SELECTOR } from '../tokens';
import { ApiActions, ApiErrorAction, ApiStartAction, ApiSuccessAction } from '../types';
import { ApiRequestEffects, buildApiHttpRequest } from './api-request.effects';

class FetchUsersStart implements ApiStartAction {
  public readonly type = '[Users] Fetch Start';
  constructor(public payload: any) {}
}

class FetchUsersSuccess<T> implements ApiSuccessAction<T> {
  public readonly type = '[Users] Fetch Success';
  constructor(public payload: T) {}
}

class FetchUsersError implements ApiErrorAction {
  public readonly type = '[Users] Fetch Error';
  constructor(public payload: HttpErrorResponse) {}
}

class FetchUsersComplete implements Action {
  public readonly type = '[Users] Fetch Complete';
}

const testActions: ApiActions<any> = {
  start: FetchUsersStart,
  success: FetchUsersSuccess,
  error: FetchUsersError,
  complete: FetchUsersComplete,
};

describe('buildApiHttpRequest', () => {
  let subject: HttpRequest<any>;

  beforeEach(() => {
    subject = buildApiHttpRequest(
      new HttpRequest('GET', 'users'),
      { apiUrl: 'http://localhost:4500', production: true },
      'the_token'
    );
  });

  it('prepends the apiUrl to url', () => {
    expect(subject.url).toEqual('http://localhost:4500/users');
  });

  it('sets auth header', () => {
    expect(subject.headers.get('Authorization')).toEqual('Bearer the_token');
  });
});

describe('ApiRequestEffects', () => {
  let httpMock: HttpTestingController;
  let actions$: ReplaySubject<any>;
  let effects: ApiRequestEffects;

  beforeEach(() => {
    TestBed.configureTestingModule({
      providers: [
        ApiRequestEffects,
        provideMockActions(() => actions$),
        { provide: APP_ENVIRONMENT, useValue: { apiUrl: 'http://localhost:3001' } },
        {
          provide: API_TOKEN_SELECTOR,
          useValue: {
            select() {
              return 'foooo';
            },
          },
        },
      ],
      imports: [HttpClientTestingModule, StoreModule.forRoot({})],
    });

    httpMock = TestBed.get(HttpTestingController);
    effects = TestBed.get(ApiRequestEffects);
    actions$ = new ReplaySubject(1);
  });

  it('should be created', () => {
    expect(effects).toBeTruthy();
  });

  describe('processApiRequest$', () => {
    describe('with complete action', () => {
      beforeEach(() => {
        actions$.next(
          new ApiRequest({ actions: testActions, request: new HttpRequest('GET', 'users') })
        );
      });

      it('fires start, success and complete on 200 request', done => {
        effects.processApiRequest$.pipe(bufferCount(3)).subscribe(actions => {
          const [start, success, complete] = actions;

          expect(start).toBeInstanceOf(FetchUsersStart);
          expect(success).toBeInstanceOf(FetchUsersSuccess);
          expect(complete).toBeInstanceOf(FetchUsersComplete);

          expect(start).toMatchSnapshot();
          expect(success).toMatchSnapshot();
          expect(complete).toMatchSnapshot();
          done();
        });

        const req = httpMock.expectOne('http://localhost:3001/users');
        expect(req.request.method).toEqual('GET');

        req.flush([]);
        httpMock.verify();
      });

      it('fires start, error and complete on 404 request', done => {
        effects.processApiRequest$.pipe(bufferCount(4)).subscribe(actions => {
          const [start, error, apiError, complete] = actions;

          expect(start).toBeInstanceOf(FetchUsersStart);
          expect(error).toBeInstanceOf(FetchUsersError);
          expect(apiError).toBeInstanceOf(ApiError);
          expect(complete).toBeInstanceOf(FetchUsersComplete);

          expect(start).toMatchSnapshot();
          expect(error).toMatchSnapshot();
          expect(apiError).toMatchSnapshot();
          expect(complete).toMatchSnapshot();
          done();
        });

        const req = httpMock.expectOne('http://localhost:3001/users');
        expect(req.request.method).toEqual('GET');

        req.flush(null, { status: 404, statusText: 'Nowhere' });
        httpMock.verify();
      });
    });

    describe('without complete action', () => {
      beforeEach(() => {
        actions$.next(
          new ApiRequest({
            actions: { start: FetchUsersStart, success: FetchUsersSuccess, error: FetchUsersError },
            request: new HttpRequest('GET', 'users'),
          })
        );
      });

      it('fires start, success and complete on 200 request', done => {
        effects.processApiRequest$.pipe(bufferCount(2)).subscribe(actions => {
          const [start, success] = actions;

          expect(start).toBeInstanceOf(FetchUsersStart);
          expect(success).toBeInstanceOf(FetchUsersSuccess);

          expect(start).toMatchSnapshot();
          expect(success).toMatchSnapshot();
          done();
        });

        const req = httpMock.expectOne('http://localhost:3001/users');
        expect(req.request.method).toEqual('GET');

        req.flush([]);
        httpMock.verify();
      });
    });
  });
});
