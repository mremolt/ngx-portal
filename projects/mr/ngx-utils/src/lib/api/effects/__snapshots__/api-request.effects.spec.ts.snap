// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiRequestEffects processApiRequest$ with complete action fires start, error and complete on 404 request 1`] = `
FetchUsersStart {
  "payload": undefined,
  "type": "[Users] Fetch Start",
}
`;

exports[`ApiRequestEffects processApiRequest$ with complete action fires start, error and complete on 404 request 2`] = `
FetchUsersError {
  "payload": HttpErrorResponse {
    "error": null,
    "headers": HttpHeaders {
      "headers": Map {},
      "lazyUpdate": null,
      "normalizedNames": Map {},
    },
    "message": "Http failure response for http://localhost:3001/users: 404 Nowhere",
    "name": "HttpErrorResponse",
    "ok": false,
    "status": 404,
    "statusText": "Nowhere",
    "url": "http://localhost:3001/users",
  },
  "type": "[Users] Fetch Error",
}
`;

exports[`ApiRequestEffects processApiRequest$ with complete action fires start, error and complete on 404 request 3`] = `
ApiError {
  "payload": HttpErrorResponse {
    "error": null,
    "headers": HttpHeaders {
      "headers": Map {},
      "lazyUpdate": null,
      "normalizedNames": Map {},
    },
    "message": "Http failure response for http://localhost:3001/users: 404 Nowhere",
    "name": "HttpErrorResponse",
    "ok": false,
    "status": 404,
    "statusText": "Nowhere",
    "url": "http://localhost:3001/users",
  },
  "type": "[API] Error",
}
`;

exports[`ApiRequestEffects processApiRequest$ with complete action fires start, error and complete on 404 request 4`] = `
FetchUsersComplete {
  "type": "[Users] Fetch Complete",
}
`;

exports[`ApiRequestEffects processApiRequest$ with complete action fires start, success and complete on 200 request 1`] = `
FetchUsersStart {
  "payload": undefined,
  "type": "[Users] Fetch Start",
}
`;

exports[`ApiRequestEffects processApiRequest$ with complete action fires start, success and complete on 200 request 2`] = `
FetchUsersSuccess {
  "payload": Array [],
  "type": "[Users] Fetch Success",
}
`;

exports[`ApiRequestEffects processApiRequest$ with complete action fires start, success and complete on 200 request 3`] = `
FetchUsersComplete {
  "type": "[Users] Fetch Complete",
}
`;

exports[`ApiRequestEffects processApiRequest$ without complete action fires start, success and complete on 200 request 1`] = `
FetchUsersStart {
  "payload": undefined,
  "type": "[Users] Fetch Start",
}
`;

exports[`ApiRequestEffects processApiRequest$ without complete action fires start, success and complete on 200 request 2`] = `
FetchUsersSuccess {
  "payload": Array [],
  "type": "[Users] Fetch Success",
}
`;
